<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Natália Martínková" />

<meta name="date" content="2024-07-16" />

<title>Importing data for genome polarisation</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Importing data for genome polarisation</h1>
<h4 class="author">Natália Martínková</h4>
<h4 class="date">2024-07-16</h4>


<div id="TOC">
<ul>
<li><a href="#quick-start" id="toc-quick-start">Quick start</a></li>
<li><a href="#principle-of-conversion-to-diem-format" id="toc-principle-of-conversion-to-diem-format">Principle of conversion
to diem format</a>
<ul>
<li><a href="#invariant-sites-and-indels" id="toc-invariant-sites-and-indels">Invariant sites and indels</a></li>
<li><a href="#informative" id="toc-informative">Sites not informative
for genome polarisation</a></li>
<li><a href="#haploid-data" id="toc-haploid-data">Haploid data</a></li>
<li><a href="#diploid-data" id="toc-diploid-data">Diploid data</a>
<ul>
<li><a href="#heterozygotes-in-dna-sequence-alignments" id="toc-heterozygotes-in-dna-sequence-alignments">Heterozygotes in DNA
sequence alignments</a></li>
<li><a href="#variant-call-format" id="toc-variant-call-format">Variant
call format</a></li>
</ul></li>
<li><a href="#polyploid-data" id="toc-polyploid-data">Polyploid
data</a></li>
</ul></li>
</ul>
</div>

<div id="quick-start" class="section level1">
<h1>Quick start</h1>
<p>The package <code>diemr</code> provides a function
<code>vcf2diem</code>, which converts diploid genotypes in a vcf file
(optionally, the vcf file can be gzipped) to a file with format required
by <code>diem</code>. To start, load the <code>diemr</code> package or
install it from CRAN if it is not yet available:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="co"># Attempt to load a package, if the package was not available, install and load</span></span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">require</span>(<span class="st">&quot;diemr&quot;</span>, <span class="at">character.only =</span> <span class="cn">TRUE</span>)){</span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a>    <span class="fu">install.packages</span>(<span class="st">&quot;diemr&quot;</span>, <span class="at">dependencies =</span> <span class="cn">TRUE</span>)</span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a>    <span class="fu">library</span>(<span class="st">&quot;diemr&quot;</span>, <span class="at">character.only =</span> <span class="cn">TRUE</span>)</span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a>}</span></code></pre></div>
<p>Next, decide whether all markers in the vcf file should be polarised
from one file, or whether the analysis will benefit from
parallelisation. Parallelisation is available on UNIX-based platforms,
and we advise to use it for large datasets. <code>diem</code> itself can
also work in parallel on multiple data <em>compartments</em>. vcf files
and <em>compartments</em> can correspond to the same thing OR you can
concatenate/split vcf2diem outputs to produce <code>diem</code> input
<em>compartments</em>.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="co"># Path to the vcf file</span></span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a>inputfile <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;myotis.vcf&quot;</span>, <span class="at">package =</span> <span class="st">&quot;diemr&quot;</span>)</span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a><span class="co"># File name for the output</span></span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a><span class="co"># If working directory does not have writing privileges, the filename must include a path to a suitable folder</span></span>
<span id="cb2-5"><a href="#cb2-5" tabindex="-1"></a>outputfile <span class="ot">&lt;-</span> <span class="st">&quot;myotis&quot;</span></span>
<span id="cb2-6"><a href="#cb2-6" tabindex="-1"></a><span class="co"># Convert the vcf file to two diem files</span></span>
<span id="cb2-7"><a href="#cb2-7" tabindex="-1"></a><span class="fu">vcf2diem</span>(<span class="at">SNP =</span> inputfile, <span class="at">filename =</span> outputfile, <span class="at">chunk =</span> <span class="dv">2</span>)</span>
<span id="cb2-8"><a href="#cb2-8" tabindex="-1"></a><span class="co"># Expecting to include 11 markers per diem file.</span></span>
<span id="cb2-9"><a href="#cb2-9" tabindex="-1"></a><span class="co"># If you expect more markers in the file, provide suitable chunk size.</span></span>
<span id="cb2-10"><a href="#cb2-10" tabindex="-1"></a><span class="co"># Done with chunk 1</span></span></code></pre></div>
<p>The <code>vcf2diem</code> function processes calls for sites into the
<code>diem</code> encoding for sites. The site order will be identical
to that in the original vcf file (but see chapter <a href="#informative">Sites not informative for genome
polarisation</a>).</p>
</div>
<div id="principle-of-conversion-to-diem-format" class="section level1">
<h1>Principle of conversion to diem format</h1>
<p>The <code>diemr</code> package uses a concise genome representation,
differentiating homozygotes for the two most frequent alleles at each
site, and their respective heterozygotes (mixed state of the two most
common alleles). All other genotypes represent an unknown state.
Specifically, the genotypes encoded as <code>0</code> represent
homozygotes for one of the two most frequent alleles, <code>1</code> are
heterozygous (mixed state) genotypes for the two most frequent alleles,
<code>2</code> are homozygotes for the other allele, and <code>U</code>
(symbol “_” is also allowed) is an unknown state that can be a missing
genotype or a genotype that includes a third (or a fourth) allele.</p>
<div id="invariant-sites-and-indels" class="section level2">
<h2>Invariant sites and indels</h2>
<p>What logically follows from this description is that genome
polarisation is meaningful only for variant sites (markers). Invariant
sites will have <strong>low support</strong> and, though they will not
change the <code>diem</code> outcome, they will slow down convergence
and obscure the nature of change between taxa. During <code>diem</code>
iterations their obscuring effect is removed by rescaling the hybrid
indices onto <span class="math inline">\([0,1]\)</span>. Including
invariant sites in genome polarisation is acceptable, and likely
unavoidable. A proportion of truly invariant sites will be mis-diagnosed
as variant due to sequencing errors.</p>
<p>At this time, genome polarisation uses variant markers where alleles
differ in nucleotide substitutions. Insertions or deletions (indels) are
not allowed to be classified among the most frequent alleles in the
<code>vcf2diem</code> function.</p>
</div>
<div id="informative" class="section level2">
<h2>Sites not informative for genome polarisation</h2>
<p>Along side invariant sites, some variant sites are also uninformative
for genome polarisation. Sites that include only missing data and
heterozygous genotypes are unpolarisable. In general applications, one
might wish to exclude sites that are only included in a <code>vcf</code>
due to the presence of a single heterozygous individual.</p>
<p>Uninformative sites will have support equal to 0, and their polarity
will thus be equal to the null polarity. Including uninformative sites
has similar consequences to including invariant sites. When a user
wishes to retain such sites, we advice to then filter polarised markers
with high support (or simply high diagnostic index, which is strongly
correlated with support) for subsequent analyses and interpretation.</p>
</div>
<div id="haploid-data" class="section level2">
<h2>Haploid data</h2>
<p>Haploid data, including markers on sex chromosomes of the
heterogametic sex, are usually shown as sequence alignments. To
illustrate the principle of genome representation used in
<code>diem</code>, let’s have an 8bp-long alignment with five
individuals.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="dv">5</span> <span class="dv">8</span></span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a>ind1   AACCTTGG</span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a>ind2   TACGATGG</span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a>ind3   TACC<span class="sc">-</span>TGT</span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a>ind4   CACGTTGG</span>
<span id="cb3-6"><a href="#cb3-6" tabindex="-1"></a>ind5   AACCTTGT</span></code></pre></div>
<p>We can see that the example alignment contains four variant markers
at sites 1, 4, 5, and 8. Marker 5 contains a deletion in ind3, while
other markers vary in different nucleotide substitutions. In marker 1,
two individuals have an allele A, two individuals have an allele T and
one has a C. This marker can be resolved in terms of two most frequent
alleles A and T, where the individual ind4 has a third allele and is
therefore considered as having an unknown genomic state. To convert the
marker into <code>diem</code> format, we must make a decision on which
allele will be encoded as 0. The choice is arbitrary. Good practice is
to <strong>keep a record of genotype encoding</strong>. Let’s decide
that all alleles in ind1 will be encoded as 0. Marker 1 will then be
<code>022_0</code>.</p>
<p>Contrary to DNA sequence alignments, diem format transposes the data.
Rows in diem input files are markers and columns represent individual
samples. The diem output for the example alignment will thus have four
lines representing the four variant markers, and five columns
representing the individuals.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a>Marker1   <span class="dv">022</span>_0</span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a>Marker4   <span class="dv">02020</span></span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a>Marker5   <span class="dv">02</span>_00</span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a>Marker8   <span class="dv">00202</span></span></code></pre></div>
<p>The diem format does not store information on marker identification
or location. This metadata must be saved separately in a ‘bed-like’ file
(with at minimum the scaffold and position for each site).
<code>diem</code> uses a prefix <code>S</code> at line starts to ensure
that all operating systems always read genotypes as strings without
attempting to convert them into numbers. Encoding of markers 4 and 8
shown above is at risk of being interpreted as numeric. The final file
in <code>diem</code> format will be:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a>S022_0</span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a>S02020</span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a>S02_00</span>
<span id="cb5-4"><a href="#cb5-4" tabindex="-1"></a>S00202</span></code></pre></div>
</div>
<div id="diploid-data" class="section level2">
<h2>Diploid data</h2>
<div id="heterozygotes-in-dna-sequence-alignments" class="section level3">
<h3>Heterozygotes in DNA sequence alignments</h3>
<p>Diploid markers in DNA sequence alignment can be resolved in an
analogous way to haploid markers. Heterozygotes in alignments can be
identified according to their <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2865858/table/T1/">IUPAC
DNA ambiguity codes</a>. For example, code W may represent an accepted
heterozygote in Marker 1, where the most frequent alleles are A and
T.</p>
</div>
<div id="variant-call-format" class="section level3">
<h3>Variant call format</h3>
<p>Variant markers in genomic context will be most often identified
using variant callers from sequence reads mapped onto a reference. Such
data are stored or can be converted to variant call format (vcf). We
implemented conversion of vcf files to diem format for the <a href="https://github.com/samtools/hts-specs/blob/master/VCFv4.2.pdf">vcf
version 4.2</a> in function <code>vcf2diem</code>. The function resolves
indels and markers with more than two alleles to obtain biallelic
genotypes for all markers included in the original vcf file.</p>
<ol style="list-style-type: decimal">
<li>Markers with the reference allele (column REF) containing insertions
are set as unknown.</li>
<li>Markers with the alternative allele (column ALT) containing only
indels are set as unknown.</li>
<li>Indels in the alternative alleles are set as unknown.</li>
<li>Overall allele counts in markers with more than two alleles
representing substitutions are calculated and the two most frequent
alleles are selected. Ties are (arbitrarily) resolved according to the
allele order in the vcf file.</li>
<li>Third and fourth most frequent substitutions are set as
unknown.</li>
</ol>
<p>The conversion might result in some markers no longer being
polymorphic. We advise checking how frequent invariant markers are after
conversion.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="co"># Import the first converted file for all individuals </span></span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a><span class="co"># and without changing marker polarity</span></span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a>myotis <span class="ot">&lt;-</span> <span class="fu">importPolarized</span>(<span class="st">&quot;myotis-01.txt&quot;</span>, </span>
<span id="cb6-4"><a href="#cb6-4" tabindex="-1"></a>                          <span class="at">changePolarity =</span> <span class="fu">rep</span>(<span class="cn">FALSE</span>, <span class="dv">11</span>), </span>
<span id="cb6-5"><a href="#cb6-5" tabindex="-1"></a>                          <span class="at">ChosenInds =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">14</span>)</span>
<span id="cb6-6"><a href="#cb6-6" tabindex="-1"></a><span class="co"># Check whether a marker includes heterozygotes </span></span>
<span id="cb6-7"><a href="#cb6-7" tabindex="-1"></a><span class="co"># or that there is more than one genotype</span></span>
<span id="cb6-8"><a href="#cb6-8" tabindex="-1"></a><span class="fu">apply</span>(myotis, <span class="at">MARGIN =</span> <span class="dv">2</span>, </span>
<span id="cb6-9"><a href="#cb6-9" tabindex="-1"></a>      <span class="at">FUN =</span> \(x) <span class="fu">any</span>(<span class="fu">grepl</span>(<span class="st">&quot;1&quot;</span>, x)) <span class="sc">|</span> <span class="fu">sum</span>(<span class="fu">levels</span>(<span class="fu">factor</span>(x)) <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">&quot;0&quot;</span>, <span class="st">&quot;1&quot;</span>, <span class="st">&quot;2&quot;</span>)) <span class="sc">&gt;</span> <span class="dv">1</span>) </span>
<span id="cb6-10"><a href="#cb6-10" tabindex="-1"></a><span class="co">#    m1    m2    m3    m4    m5    m6    m7    m8    m9   m10   m11 </span></span>
<span id="cb6-11"><a href="#cb6-11" tabindex="-1"></a><span class="co"># FALSE FALSE FALSE FALSE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE </span></span>
<span id="cb6-12"><a href="#cb6-12" tabindex="-1"></a>         </span></code></pre></div>
<p>Starting from version 1.2.2, <code>vcf2diem</code> removes markers
that include</p>
<ol style="list-style-type: decimal">
<li>only missing data and heterozygotes,</li>
<li>missing data, homozygotes for the most frequent allele and one
heterozygote,</li>
<li>optionally missing data, homozygotes for one allele and
heterozygotes.</li>
</ol>
<p>We include a list of the included markers in a bed-like file ending
with <em>-includedSites.txt</em> in the working directory. The file
includes information from columns <code>CHROM</code>, <code>POS</code>,
and <code>QUAL</code> for the respective markers, and records the (good
practice) <strong>record of genotype encoding</strong>. The file
includedLoci.txt must be checked when interpreting polarised markers to
ensure correct marker annotation.</p>
<p>We include a list of the removed markers in a bed-like file ending
with <em>-omittedSites.txt</em> in the working directory. The file
includes information from columns <code>CHROM</code>, <code>POS</code>,
and <code>QUAL</code> for the respective markers, and provides the
reason, why the marker was omitted.</p>
</div>
</div>
<div id="polyploid-data" class="section level2">
<h2>Polyploid data</h2>
<p>The same principles apply to higher ploidy datasets. Preparing data
in <code>diem</code> format for genome polarisation includes these
steps:</p>
<ol style="list-style-type: decimal">
<li>Remove indels.</li>
<li>Identify two most frequent alleles for each site.</li>
<li>Encode homozygous and heterozygous (mixed) genotypes that include
only the two most frequent alleles.</li>
</ol>
<p>Note that <code>vcf2diem</code> only converts diploid datasets at
this time, and should not be used to convert polyploid vcf files.</p>
</div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
